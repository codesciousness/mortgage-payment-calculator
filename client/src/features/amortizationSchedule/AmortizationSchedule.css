.AmortizationSchedule {
    text-align: left;
}

.AmortizationSchedule__label {
    font-size: 0.75rem;
    margin-bottom: 0;
    color: Gray;
}

.AmortizationSchedule__amount {
    font-size: 1.75rem;
    font-weight: bold;
    margin-top: 0;
    position: relative;
}

.AmortizationSchedule__amount:before {
    content: '$';
    position: absolute;
    left: -10px;
    top: 3px;
    font-size: 1rem;
    font-weight: bold;
}

.AmortizationSchedule__LineChart__container {
    display: flex;
    align-items: center;
    width: 100%;
}

.AmortizationSchedule__LineChart {
    width: 100%;
}