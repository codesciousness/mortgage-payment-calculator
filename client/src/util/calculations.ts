export const isInt = (num: number | string): boolean => {
    let number: number;
    if (typeof num === 'string') number = stringToNum(num)
    else number = num;
    return Number.isInteger(number);
};

export const rmvCommas = (num: string): string => {
    return num.replaceAll(',', '');
};

export const addCommas = (num: number | string): string => {
    let numString: string;
    if (typeof num === 'number') numString = numToString(num)
    else numString = rmvCommas(num);
    return numString.replace(/\B(?<!\.\d*)(?=(\d{3})+(?!\d))/g, ",");
};

export const stringToNum = (str: string): number => {
    const num = parseFloat(rmvCommas(str));
    if (Number.isNaN(num)) return 0;
    return num;
};

export const numToString = (num: number): string => {
    return `${num}`;
};

export const round = (num: number | string): string => {
    let number: number;
    if (num === '') return '';
    if (typeof num === 'string') number = stringToNum(num)
    else number = num;
    return numToString(Math.round(number));
};

export const fixDecimal = (num: number | string): string => {
    let numString: string;
    let decimals: string;
    if (typeof num === 'number') numString = numToString(num)
    else numString = rmvCommas(num);
    decimals = numString.split('.')[1];
    if (decimals && decimals.length > 2) return numString.slice(0, numString.length - 1)
    else return numString;
};

export const formatAmount = (num: number | string): string => {
    return addCommas(round(num));
};

export const formatPercent = (num: number | string): string => {
    return addCommas(fixDecimal(num));
};

export const fromPercent = (percent: string, total: string): string => {
    const percentNum = stringToNum(percent)/100;
    const totalNum = stringToNum(total);
    const amount = percentNum * totalNum;
    return formatAmount(amount);
};

export const toPercent = (amount: string, total: string): string => {
    const amountNum = stringToNum(amount);
    const totalNum = stringToNum(total);
    const percent = (amountNum / totalNum) * 100;
    if (isInt(percent)) {
        return addCommas(percent);
    }
    return addCommas(percent.toFixed(2));
};

export const mortgageFormula = (loanAmount: string, interestRate: number, loanTerm: number): string => {
    const principal = stringToNum(loanAmount);
    const r = interestRate/12;
    const n = loanTerm * 12
    const monthlyPayment = principal * (r*(1 + r)**n)/((1 + r**n) - 1);
    return numToString(monthlyPayment);
};