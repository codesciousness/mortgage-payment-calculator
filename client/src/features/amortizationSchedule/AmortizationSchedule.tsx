import React from 'react';
import AmortizationTable from '../amortizationTable/AmortizationTable';
import LineChart from '../lineChart/LineChart';
import LineChartLegend from '../lineChartLegend/LineChartLegend';
import Box from '@mui/material/Box';
import Divider from '@mui/material/Divider';
import './AmortizationSchedule.css';
import { selectLoanAmount, selectTotalInterest, selectLoanCost, selectPayoffDate } from '../loansSlice';
import { useAppSelector } from '../../app/hooks';

const AmortizationSchedule = (): JSX.Element => {
    const loanAmount = useAppSelector(selectLoanAmount);
    const totalInterest = useAppSelector(selectTotalInterest);
    const loanCost = useAppSelector(selectLoanCost);
    const payoffDate = useAppSelector(selectPayoffDate);

    return (
        <section id='AmortizationSchedule' className='AmortizationSchedule'>
            <h2 className='AmortizationSchedule__title'>Amortization for mortgage loan</h2>
            <p className='AmortizationSchedule__content'>Amortization is paying off debt over time in equal installments. As the term of your mortgage loan progresses,
                a larger share of your payment goes toward paying down the principal until the loan is paid in full at the end of your term.</p>
            <Box className='AmortizationSchedule__loanTotals' sx={{ borderTop: 1, borderBottom: 1, borderColor: 'divider', }}>
                <div className='AmortizationSchedule__container'>
                    <p className='AmortizationSchedule__label'>Loan amount</p>
                    <p className='AmortizationSchedule__amount loan'>{loanAmount}</p>
                </div>
                <div className='AmortizationSchedule__container'>
                    <p className='AmortizationSchedule__label'>Total interest paid</p>
                    <p className='AmortizationSchedule__amount interest'>{totalInterest}</p>
                </div>
                <div className='AmortizationSchedule__container'>
                    <p className='AmortizationSchedule__label'>Total cost of loan</p>
                    <p className='AmortizationSchedule__amount cost'>{loanCost}</p>
                </div>
                <div className='AmortizationSchedule__container'>
                    <p className='AmortizationSchedule__label'>Payoff Date</p>
                    <p className='AmortizationSchedule__date'>{payoffDate}</p>
                </div>
            </Box>
            <div className='AmortizationSchedule__LineChart__container'>
                <div className='AmortizationSchedule__LineChart'>
                    <LineChart/>
                </div>
                <LineChartLegend/>
            </div>
            <Divider className='AmortizationSchedule__divider' sx={{ margin: '2rem' }}/>
            <AmortizationTable />
        </section>
    );
};

export default AmortizationSchedule;