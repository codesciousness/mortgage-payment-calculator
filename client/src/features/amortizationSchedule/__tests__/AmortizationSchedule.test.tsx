import React from 'react';
import { Provider } from 'react-redux';
import { render, screen } from '@testing-library/react';
import { create } from 'react-test-renderer';
import { store } from '../../../app/store';
import AmortizationSchedule from '../AmortizationSchedule';
import { dateToString } from '../../../util/calculations';

jest.mock('../../lineChart/LineChart', () => () => (<section data-testid="LineChart">Line Chart</section>));
jest.mock('../../lineChartLegend/LineChartLegend', () => () => (<section data-testid="LineChartLegend">Line Chart Legend</section>));
jest.mock('../../amortizationTable/AmortizationTable', () => () => (<section data-testid="AmortizationTable">Amortization Table</section>));

describe('<AmortizationSchedule />', () => {
    test('AmortizationSchedule component is defined', () => {
        expect(<AmortizationSchedule/>).toBeDefined();
    });
    test('component matches the snapshot', () => {
        const component = create(
            <Provider store={store}>
                <AmortizationSchedule/>
            </Provider>
        );
        expect(component.toJSON()).toMatchSnapshot();
    });
    test('renders correctly with initial state from Redux store', () => {
        render(
            <Provider store={store}>
                <AmortizationSchedule/>
            </Provider>
        );
        const date = new Date();
        const year = date.getFullYear();
        const year30 = year + 30;
        date.setFullYear(year30);

        expect(screen.getByText('Amortization for mortgage loan')).toBeInTheDocument();
        expect(screen.getByText('Loan amount')).toBeInTheDocument();
        expect(screen.getByText('315,000')).toBeInTheDocument();
        expect(screen.getByText('Total interest paid')).toBeInTheDocument();
        expect(screen.getByText('346,772')).toBeInTheDocument();
        expect(screen.getByText('Total cost of loan')).toBeInTheDocument();
        expect(screen.getByText('661,772')).toBeInTheDocument();
        expect(screen.getByText('Payoff Date')).toBeInTheDocument();
        expect(screen.getByText(dateToString(date))).toBeInTheDocument();
    });
    test('renders a LineChart component', () => {
        render(
            <Provider store={store}>
                <AmortizationSchedule/>
            </Provider>
        );
        expect(screen.getByTestId('LineChart')).toBeInTheDocument();
        expect(screen.getByText('Line Chart')).toBeInTheDocument();
    });
    test('renders a LineChartLegend component', () => {
        render(
            <Provider store={store}>
                <AmortizationSchedule/>
            </Provider>
        );
        expect(screen.getByTestId('LineChartLegend')).toBeInTheDocument();
        expect(screen.getByText('Line Chart Legend')).toBeInTheDocument();
    });
    test('renders an AmortizationTable component', () => {
        render(
            <Provider store={store}>
                <AmortizationSchedule/>
            </Provider>
        );
        expect(screen.getByTestId('AmortizationTable')).toBeInTheDocument();
        expect(screen.getByText('Amortization Table')).toBeInTheDocument();
    });
});