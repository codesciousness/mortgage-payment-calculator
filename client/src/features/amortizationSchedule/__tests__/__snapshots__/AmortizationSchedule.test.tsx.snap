// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AmortizationSchedule /> component matches the snapshot 1`] = `
<section
  className="AmortizationSchedule"
  id="AmortizationSchedule"
>
  <h2
    className="AmortizationSchedule__title"
  >
    Amortization for mortgage loan
  </h2>
  <p
    className="AmortizationSchedule__content"
  >
    Amortization is paying off debt over time in equal installments. As the term of your mortgage loan progresses, a larger share of your payment goes toward paying down the principal until the loan is paid in full at the end of your term.
  </p>
  <div
    className="AmortizationSchedule__loanTotals MuiBox-root css-1sd9iv3"
  >
    <div
      className="AmortizationSchedule__container"
    >
      <p
        className="AmortizationSchedule__label"
      >
        Loan amount
      </p>
      <p
        className="AmortizationSchedule__amount loan"
      >
        <span
          className="AmortizationSchedule__sign"
        >
          $
        </span>
        315,000
      </p>
    </div>
    <div
      className="AmortizationSchedule__container"
    >
      <p
        className="AmortizationSchedule__label"
      >
        Total interest paid
      </p>
      <p
        className="AmortizationSchedule__amount interest"
      >
        <span
          className="AmortizationSchedule__sign"
        >
          $
        </span>
        346,772
      </p>
    </div>
    <div
      className="AmortizationSchedule__container"
    >
      <p
        className="AmortizationSchedule__label"
      >
        Total cost of loan
      </p>
      <p
        className="AmortizationSchedule__amount cost"
      >
        <span
          className="AmortizationSchedule__sign"
        >
          $
        </span>
        661,772
      </p>
    </div>
    <div
      className="AmortizationSchedule__container"
    >
      <p
        className="AmortizationSchedule__label"
      >
        Payoff Date
      </p>
      <p
        className="AmortizationSchedule__date"
      >
        05/2053
      </p>
    </div>
  </div>
  <div
    className="AmortizationSchedule__LineChart__container"
  >
    <div
      className="AmortizationSchedule__LineChart"
    >
      <section
        data-testid="LineChart"
      >
        Line Chart
      </section>
    </div>
    <section
      data-testid="LineChartLegend"
    >
      Line Chart Legend
    </section>
  </div>
  <hr
    className="MuiDivider-root MuiDivider-fullWidth AmortizationSchedule__divider css-lmcth4-MuiDivider-root"
  />
  <section
    data-testid="AmortizationTable"
  >
    Amortization Table
  </section>
</section>
`;
